[WGSL] Cache parameterized types

tadeuzagallo opened the pull request:
#### 5ebfd986c0ce5db75eb4bed79902f577b3a4f3d7
<pre>
[WGSL] Cache parameterized types
<a href="https://bugs.webkit.org/show_bug.cgi?id=252326">https://bugs.webkit.org/show_bug.cgi?id=252326</a>
rdar://105503636

Reviewed by Myles C. Maxfield.

Hash cons parameterized types so they can be checked for equality
with pointer identity.

* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::VectorKey::extra const):
(WGSL::MatrixKey::extra const):
(WGSL::ArrayKey::extra const):
(WGSL::TypeStore::TypeCache::find const):
(WGSL::TypeStore::TypeCache::insert):
(WGSL::TypeStore::constructType):
(WGSL::TypeStore::arrayType):
(WGSL::TypeStore::vectorType):
(WGSL::TypeStore::matrixType):
(WGSL::TypeStore::allocateConstructor):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):

Canonical link: <a href="https://commits.webkit.org/260358@main">https://commits.webkit.org/260358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/daabd2f87b6fcef209c4fd541308eccb7ffcfac3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108102 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117227 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8468 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100306 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113870 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97196 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41901 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28834 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10054 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30182 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10776 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49773 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7167 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12357 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WGSL/TypeStore.cpp:
```diff
@@ -34,6 +34,44 @@ namespace WGSL {
 
 using namespace Types;
 
+struct VectorKey {
```
Do we need these keys? Why not implement a hasher for `Type`?

tadeuzagallo replied:
We're trying to map e.g. `T + 2` to `vec2<T>`, so just hashing the type wouldn't give us enough information.

djg commented on Source/WebGPU/WGSL/TypeStore.h:
```diff
@@ -60,6 +61,18 @@ class TypeStore {
+        void insert(const Key&, Type*);
+
+    private:
+        HashMap<std::tuple<Type*, uint64_t>, Type*> m_storage;
```
Mapping a `Type*` + extra magic number to `Type*` seems a bit strange to me. Is this how it's done in JSC?

tadeuzagallo replied:
I don't know if we do that specifically in JSC, but it's pretty common to serialize data into unsigned types. I think this gives us enough information to put all the types in the same HashMap, avoid the overhead of having one HashMap for each kind of type.

tadeuzagallo replied:
I'll add a comment here to explain my thoughts and the layout of the second word

djg commented on Source/WebGPU/WGSL/TypeStore.cpp:
```diff
@@ -67,27 +105,41 @@ Type* TypeStore::structType(const AST::Identifier& name)
+    Type* type = m_cache.find(key);
+    if (type)
+        return type;
+    type = allocateType<Matrix>(elementType, columns, rows);
```
Shouldn't we just preallocate all the matrix types? There's only a limit number of vector and matrix types.

tadeuzagallo replied:
We could, but that'd require 27 matrix types ahead of time, which seems unnecessary? I think very few programs will use them all

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Cache parameterized types
https://bugs.webkit.org/show_bug.cgi?id=252326
rdar://105503636

Reviewed by Myles C. Maxfield.

Hash cons parameterized types so they can be checked for equality
with pointer identity.

* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::VectorKey::extra const):
(WGSL::MatrixKey::extra const):
(WGSL::ArrayKey::extra const):
(WGSL::TypeStore::TypeCache::find const):
(WGSL::TypeStore::TypeCache::insert):
(WGSL::TypeStore::constructType):
(WGSL::TypeStore::arrayType):
(WGSL::TypeStore::vectorType):
(WGSL::TypeStore::matrixType):
(WGSL::TypeStore::allocateConstructor):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):

Canonical link: https://commits.webkit.org/260358@main
